RFC.md

```diff
@@ -1,12 +1,234 @@
+if the _flattened representation of the indeces alias_. So, in this case,
+pick the parameter values:
+
+```
+n = m = 3
+x1 = 1, y1 = 0; B[x1][y1] = nx1+y1 = 3*1+0=3
+x2 = 0, y2 = 3; B[x2][y2] = nx2+y2 = 3*0+3=3
+```
+
+Hence, the arrays _do alias_, and the transformation proposed in `ex1_opt`
+is unsound in general.
+
+However, there are languages that provide this "tuple-indexed" semantics,
+where one can infer that the indexing:
+`[x1][y1] != [x2][y2]` iff `x1 != x2 || y1 != y2`.
+
+Julia, Fortran, and Chapel are examples of such languages which target
+LLVM.
+
+Currently, the LLVM semantics of `getelementptr` talk purely of
+the C style flattened views of arrays. This inhibits the ability of the optimiser
+to understand the multidimensional examples as given above, and we
+are forced to make conservative assumptions, inhibiting optimisation. 
+This information must ideally be expressible in LLVM, such that LLVM's optimisers and
+alias analysis can use this information to model multidimensional-array
+semantics.
+
+
+
+## Evaluation of the impact of the intrinsic on accuracy of dependence analysis
+
+- This has been implemented in an exprimental branch of Polly, and was used 
+on the COSMO climate weather model. This greatly helped increase the accuracy
+of Polly's analysis, since we eliminated the guessing game from the array analysis.
+
+- This has also been implemented as part of a GSoC effort to unify
+Chapel and Polly. 
+
+- Molly, the distributed version of Polly written by Michael Kruse for his PhD
+  also implemented a similar scheme. In his use-case, guessing as we perform
+  optimistically was not possible, so this kind of intrinsic was _necessary_
+  for the application, not just _good to have_. More details are available
+  in his PhD thesis: [Lattice QCD Optimization and Polytopic Representations of Distributed Memory](https://tel.archives-ouvertes.fr/tel-01078440).
+  In particular, Chapter 9 contains a detailed discussion.
+
+# Representations
+
+## Instruction / Intrinsic
+
+### Syntax
+```
+<result> = multidim_array_index <ty> <ty>* <ptrval> {<stride>, <idx>}*
+```
+
+### Overview:
+The `multidim_array_index` instruction is used to get the address of 
+an element from an array. It performs address calcuation only and 
+does not access memory. It is similar to `getelementptr`. However, it imposes
+additional semantics which allows the optimiser to provide better optimisations
+than `getlementptr`.
+
+
+
+### Arguments:
+The first argument is always a type used as the basis for the calculations. The
+second argument is always a pointer, and is the base address to start the
+calculation from. The remaining arguments are a list of pairs. Each pair
+contains a dimension stride, and an offset with respect to that stride.
+
+
+### Semantics:
+
+`multidim_array_index` represents a multi-dimensional array index, In particular, this will
+mean that we will assume that all indeces `<idx_n>` are non-negative.
```

Comment:
indeces -> indices

I think you also need to say here that the optimizer can assume that the multidim_array_index calls allow the optimizer to assume non-aliasing. I don't think you get this solely from the math. You also need the restriction that the math does not "overflow".

I would say this after what you have so far:

Additionally, we assume that, for each `<str_i>`, `<idx_i>` pair, that `0 <= idx_i < str_i`.

Optimizations can assume that, given two `multidim_array_index` instructions with matching types:
  * `multidim_array_index <ty> <ty>* <ptrvalA> <strA_1>, <idxA_1>, ..., <strA_N>, <idxA_N>`
  * `multidim_array_index <ty> <ty>* <ptrvalB> <strB_1>, <idxB_1>, ..., <strB_N>, <idxb_N>`

If `ptrvalA == ptrvalB` and the strides are equal (`strA_1 == strB_1 && ... && strA_N == strB_N`) then:
 * if the indices are equal (that is,`idxA_1 == idxB_1 && ... && idxA_N == idxB_N`) then the resulting pointer values are equal
 * if any index value is not equal (that is, for some i, `idxA_i != idxB_i`) then the resulting pointer values are not equal